File: swift/src/Branch.swift
/// MuonGit - First-class branch API
import Foundation

public enum BranchType: Sendable {
    case local
    case remote
}

public struct BranchUpstream: Sendable, Equatable {
    public let remoteName: String
    public let mergeRef: String

    public init(remoteName: String, mergeRef: String) {
        self.remoteName = remoteName
        self.mergeRef = mergeRef
    }
}

public struct Branch: Sendable, Equatable {
    public let name: String
    public let referenceName: String
    public let target: OID?
    public let kind: BranchType
    public let isHEAD: Bool
    public let upstream: BranchUpstream?
}

public func createBranch(
    gitDir: String,
    name: String,
    target: OID? = nil,
    force: Bool = false
) throws -> Branch {
    let refName = localBranchRef(name)
    let refdb = RefDb(gitDir: gitDir)

    if referenceExists(gitDir: gitDir, name: refName) {
        if !force {
            throw MuonGitError.conflict("branch '\(name)' already exists")
        }
        _ = try refdb.delete(name: refName)
    }

    let targetOID = try target ?? headTargetOID(gitDir: gitDir)
    try refdb.write(name: refName, oid: targetOID)
    return try lookupBranch(gitDir: gitDir, name: name, kind: .local)
}

public func lookupBranch(gitDir: String, name: String, kind: BranchType) throws -> Branch {
    try buildBranch(gitDir: gitDir, refName: branchRefName(name: name, kind: kind), kind: kind)
}

public func listBranches(gitDir: String, kind: BranchType? = nil) throws -> [Branch] {
    let refs = try listReferences(gitDir: gitDir)
    var branches: [Branch] = []

    for (refName, _) in refs {
        if let branchType = branchType(for: refName) {
            if kind == nil || kind == branchType {
                branches.append(try buildBranch(gitDir: gitDir, refName: refName, kind: branchType))
            }
        }
    }

    return branches.sorted { $0.referenceName < $1.referenceName }
}

public func renameBranch(
    gitDir: String,
    oldName: String,
    newName: String,
    force: Bool = false
) throws -> Branch {
    let oldRef = localBranchRef(oldName)
    let newRef = localBranchRef(newName)
    if oldRef == newRef {
        return try lookupBranch(gitDir: gitDir, name: newName, kind: .local)
    }

    let refdb = RefDb(gitDir: gitDir)
    let oldBranch = try refdb.read(name: oldRef)

    if referenceExists(gitDir: gitDir, name: newRef) {
        if !force {
            throw MuonGitError.conflict("branch '\(newName)' already exists")
        }
        if try currentHEADRef(gitDir: gitDir) == newRef {
            throw MuonGitError.conflict("cannot replace checked out branch '\(newName)'")
        }
        _ = try deleteBranch(gitDir: gitDir, name: newName, kind: .local)
    }

    if let symbolicTarget = oldBranch.symbolicTarget {
        try refdb.writeSymbolic(name: newRef, target: symbolicTarget)
    } else if let target = oldBranch.target {
        try refdb.write(name: newRef, oid: target)
    } else {
        throw MuonGitError.invalidObject("branch '\(oldName)' has no target")
    }
    _ = try refdb.delete(name: oldRef)
    try moveBranchUpstream(gitDir: gitDir, oldName: oldName, newName: newName)

    if try currentHEADRef(gitDir: gitDir) == oldRef {
        try refdb.writeSymbolic(name: "HEAD", target: newRef)
    }

    return try lookupBranch(gitDir: gitDir, name: newName, kind: .local)
}

@discardableResult
public func deleteBranch(gitDir: String, name: String, kind: BranchType) throws -> Bool {
    let refName = branchRefName(name: name, kind: kind)
    let headRef = try currentHEADRef(gitDir: gitDir)
    if kind == .local && headRef == refName {
        throw MuonGitError.conflict("cannot delete checked out branch '\(name)'")
    }

    let deleted = try RefDb(gitDir: gitDir).delete(name: refName)
    if deleted && kind == .local {
        try clearBranchUpstream(gitDir: gitDir, name: name)
    }
    return deleted
}

public func branchUpstream(gitDir: String, name: String) throws -> BranchUpstream? {
    let config = try loadRepoConfig(gitDir: gitDir)
    let section = branchSection(name)
    let remote = config.get(section: section, key: "remote")
    let merge = config.get(section: section, key: "merge")

    switch (remote, merge) {
    case let (remoteName?, mergeRef?):
        return BranchUpstream(remoteName: remoteName, mergeRef: mergeRef)
    case (nil, nil):
        return nil
    default:
        throw MuonGitError.invalidSpec("branch '\(name)' has incomplete upstream config")
    }
}

public func setBranchUpstream(gitDir: String, name: String, upstream: BranchUpstream?) throws {
    let refName = localBranchRef(name)
    guard referenceExists(gitDir: gitDir, name: refName) else {
        throw MuonGitError.notFound("branch '\(name)' not found")
    }

    let config = try loadRepoConfig(gitDir: gitDir)
    let section = branchSection(name)
    if let upstream {
        config.set(section: section, key: "remote", value: upstream.remoteName)
        config.set(section: section, key: "merge", value: upstream.mergeRef)
    } else {
        config.unset(section: section, key: "remote")
        config.unset(section: section, key: "merge")
    }
    try config.save()
}

public extension Repository {
    func createBranch(name: String, target: OID? = nil, force: Bool = false) throws -> Branch {
        try MuonGit.createBranch(gitDir: gitDir, name: name, target: target, force: force)
    }

    func lookupBranch(name: String, kind: BranchType) throws -> Branch {
        try MuonGit.lookupBranch(gitDir: gitDir, name: name, kind: kind)
    }

    func listBranches(kind: BranchType? = nil) throws -> [Branch] {
        try MuonGit.listBranches(gitDir: gitDir, kind: kind)
    }
}

private func buildBranch(gitDir: String, refName: String, kind: BranchType) throws -> Branch {
    let reference = try RefDb(gitDir: gitDir).read(name: refName)
    let target = reference.isSymbolic ? (try? resolveReference(gitDir: gitDir, name: refName)) : reference.target
    guard let shortName = shortBranchName(refName: refName, kind: kind) else {
        throw MuonGitError.invalidSpec("not a branch reference: \(refName)")
    }

    return Branch(
        name: shortName,
        referenceName: refName,
        target: target,
        kind: kind,
        isHEAD: try currentHEADRef(gitDir: gitDir) == refName,
        upstream: kind == .local ? try branchUpstream(gitDir: gitDir, name: shortName) : nil
    )
}

private func branchRefName(name: String, kind: BranchType) -> String {
    switch kind {
    case .local: return localBranchRef(name)
    case .remote: return "refs/remotes/\(name)"
    }
}

private func localBranchRef(_ name: String) -> String {
    "refs/heads/\(name)"
}

private func shortBranchName(refName: String, kind: BranchType) -> String? {
    switch kind {
    case .local:
        return refName.hasPrefix("refs/heads/") ? String(refName.dropFirst("refs/heads/".count)) : nil
    case .remote:
        return refName.hasPrefix("refs/remotes/") ? String(refName.dropFirst("refs/remotes/".count)) : nil
    }
}

private func branchType(for refName: String) -> BranchType? {
    if refName.hasPrefix("refs/heads/") {
        return .local
    }
    if refName.hasPrefix("refs/remotes/") {
        return .remote
    }
    return nil
}

private func currentHEADRef(gitDir: String) throws -> String? {
    let head = try readReference(gitDir: gitDir, name: "HEAD")
    guard head.hasPrefix("ref: ") else { return nil }
    let target = String(head.dropFirst(5)).trimmingCharacters(in: .whitespacesAndNewlines)
    return target.hasPrefix("refs/heads/") ? target : nil
}

private func headTargetOID(gitDir: String) throws -> OID {
    let head = try readReference(gitDir: gitDir, name: "HEAD")
    if head.hasPrefix("ref: ") {
        do {
            return try resolveReference(gitDir: gitDir, name: "HEAD")
        } catch MuonGitError.notFound(_) {
            throw MuonGitError.unbornBranch
        }
    }
    return OID(hex: head.trimmingCharacters(in: .whitespacesAndNewlines))
}

private func branchSection(_ name: String) -> String {
    "branch.\(name)"
}

private func loadRepoConfig(gitDir: String) throws -> Config {
    let configPath = (gitDir as NSString).appendingPathComponent("config")
    if FileManager.default.fileExists(atPath: configPath) {
        return try Config.load(from: configPath)
    }
    return Config(path: configPath)
}

private func clearBranchUpstream(gitDir: String, name: String) throws {
    let configPath = (gitDir as NSString).appendingPathComponent("config")
    guard FileManager.default.fileExists(atPath: configPath) else {
        return
    }
    let config = try Config.load(from: configPath)
    let section = branchSection(name)
    config.unset(section: section, key: "remote")
    config.unset(section: section, key: "merge")
    try config.save()
}

private func moveBranchUpstream(gitDir: String, oldName: String, newName: String) throws {
    let upstream = try branchUpstream(gitDir: gitDir, name: oldName)
    try clearBranchUpstream(gitDir: gitDir, name: oldName)
    if let upstream {
        try setBranchUpstream(gitDir: gitDir, name: newName, upstream: upstream)
    }
}

private func referenceExists(gitDir: String, name: String) -> Bool {
    (try? readReference(gitDir: gitDir, name: name)) != nil
}

File: rust/src/branch.rs
//! First-class branch API built on top of refs/refdb.

use std::path::Path;

use crate::config::Config;
use crate::error::MuonGitError;
use crate::oid::OID;
use crate::refdb::RefDb;
use crate::repository::Repository;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BranchType {
    Local,
    Remote,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BranchUpstream {
    pub remote_name: String,
    pub merge_ref: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Branch {
    pub name: String,
    pub reference_name: String,
    pub target: Option<OID>,
    pub kind: BranchType,
    pub is_head: bool,
    pub upstream: Option<BranchUpstream>,
}

pub fn create_branch(
    git_dir: &Path,
    name: &str,
    target: Option<&OID>,
    force: bool,
) -> Result<Branch, MuonGitError> {
    let ref_name = local_branch_ref(name);
    let refdb = RefDb::open(git_dir);

    if reference_exists(git_dir, &ref_name) {
        if !force {
            return Err(MuonGitError::Conflict(format!(
                "branch '{}' already exists",
                name
            )));
        }
        refdb.delete(&ref_name)?;
    }

    let target_oid = match target {
        Some(oid) => oid.clone(),
        None => head_target_oid(git_dir)?,
    };
    refdb.write(&ref_name, &target_oid)?;
    lookup_branch(git_dir, name, BranchType::Local)
}

pub fn lookup_branch(
    git_dir: &Path,
    name: &str,
    kind: BranchType,
) -> Result<Branch, MuonGitError> {
    let ref_name = branch_ref_name(name, kind);
    build_branch(git_dir, &ref_name, kind)
}

pub fn list_branches(
    git_dir: &Path,
    kind: Option<BranchType>,
) -> Result<Vec<Branch>, MuonGitError> {
    let refs = crate::refs::list_references(git_dir)?;
    let mut branches = Vec::new();

    for (ref_name, _) in refs {
        if let Some((branch_name, branch_type)) = branch_name_and_type(&ref_name) {
            if kind.is_none() || kind == Some(branch_type) {
                branches.push(build_branch(git_dir, &ref_name, branch_type)?);
                if branches.last().is_some() {
                    let _ = &branch_name;
                }
            }
        }
    }

    branches.sort_by(|a, b| a.reference_name.cmp(&b.reference_name));
    Ok(branches)
}

pub fn rename_branch(
    git_dir: &Path,
    old_name: &str,
    new_name: &str,
    force: bool,
) -> Result<Branch, MuonGitError> {
    let old_ref = local_branch_ref(old_name);
    let new_ref = local_branch_ref(new_name);
    if old_ref == new_ref {
        return lookup_branch(git_dir, new_name, BranchType::Local);
    }

    let refdb = RefDb::open(git_dir);
    let old_branch = refdb.read(&old_ref)?;
    if reference_exists(git_dir, &new_ref) {
        if !force {
            return Err(MuonGitError::Conflict(format!(
                "branch '{}' already exists",
                new_name
            )));
        }
        if current_head_ref(git_dir)?.as_deref() == Some(new_ref.as_str()) {
            return Err(MuonGitError::Conflict(format!(
                "cannot replace checked out branch '{}'",
                new_name
            )));
        }
        delete_branch(git_dir, new_name, BranchType::Local)?;
    }

    if let Some(symbolic_target) = old_branch.symbolic_target {
        refdb.write_symbolic(&new_ref, &symbolic_target)?;
    } else if let Some(target) = old_branch.target {
        refdb.write(&new_ref, &target)?;
    } else {
        return Err(MuonGitError::InvalidObject(format!(
            "branch '{}' has no target",
            old_name
        )));
    }
    refdb.delete(&old_ref)?;
    move_branch_upstream(git_dir, old_name, new_name)?;

    if current_head_ref(git_dir)?.as_deref() == Some(old_ref.as_str()) {
        refdb.write_symbolic("HEAD", &new_ref)?;
    }

    lookup_branch(git_dir, new_name, BranchType::Local)
}

pub fn delete_branch(
    git_dir: &Path,
    name: &str,
    kind: BranchType,
) -> Result<bool, MuonGitError> {
    let ref_name = branch_ref_name(name, kind);
    if kind == BranchType::Local && current_head_ref(git_dir)?.as_deref() == Some(ref_name.as_str()) {
        return Err(MuonGitError::Conflict(format!(
            "cannot delete checked out branch '{}'",
            name
        )));
    }

    let deleted = RefDb::open(git_dir).delete(&ref_name)?;
    if deleted && kind == BranchType::Local {
        clear_branch_upstream(git_dir, name)?;
    }
    Ok(deleted)
}

pub fn branch_upstream(git_dir: &Path, name: &str) -> Result<Option<BranchUpstream>, MuonGitError> {
    let config = load_repo_config(git_dir)?;
    let section = branch_section(name);
    let remote = config.get(&section, "remote").map(|value| value.to_string());
    let merge = config.get(&section, "merge").map(|value| value.to_string());

    match (remote, merge) {
        (Some(remote_name), Some(merge_ref)) => Ok(Some(BranchUpstream {
            remote_name,
            merge_ref,
        })),
        (None, None) => Ok(None),
        _ => Err(MuonGitError::InvalidSpec(format!(
            "branch '{}' has incomplete upstream config",
            name
        ))),
    }
}

pub fn set_branch_upstream(
    git_dir: &Path,
    name: &str,
    upstream: Option<BranchUpstream>,
) -> Result<(), MuonGitError> {
    let branch_ref = local_branch_ref(name);
    if !reference_exists(git_dir, &branch_ref) {
        return Err(MuonGitError::NotFound(format!(
            "branch '{}' not found",
            name
        )));
    }

    let config_path = git_dir.join("config");
    let mut config = if config_path.exists() {
        Config::load(&config_path)?
    } else {
        Config::with_path(&config_path)
    };
    let section = branch_section(name);

    if let Some(upstream) = upstream {
        config.set(&section, "remote", &upstream.remote_name);
        config.set(&section, "merge", &upstream.merge_ref);
    } else {
        config.unset(&section, "remote");
        config.unset(&section, "merge");
    }
    config.save()
}

impl Repository {
    pub fn create_branch(
        &self,
        name: &str,
        target: Option<&OID>,
        force: bool,
    ) -> Result<Branch, MuonGitError> {
        create_branch(self.git_dir(), name, target, force)
    }

    pub fn lookup_branch(
        &self,
        name: &str,
        kind: BranchType,
    ) -> Result<Branch, MuonGitError> {
        lookup_branch(self.git_dir(), name, kind)
    }

    pub fn list_branches(
        &self,
        kind: Option<BranchType>,
    ) -> Result<Vec<Branch>, MuonGitError> {
        list_branches(self.git_dir(), kind)
    }
}

fn build_branch(git_dir: &Path, ref_name: &str, kind: BranchType) -> Result<Branch, MuonGitError> {
    let reference = RefDb::open(git_dir).read(ref_name)?;
    let target = if reference.is_symbolic() {
        crate::refs::resolve_reference(git_dir, ref_name).ok()
    } else {
        reference.target.clone()
    };
    let name = short_branch_name(ref_name, kind).ok_or_else(|| {
        MuonGitError::InvalidSpec(format!("not a branch reference: {}", ref_name))
    })?;

    Ok(Branch {
        name: name.to_string(),
        reference_name: ref_name.to_string(),
        target,
        kind,
        is_head: current_head_ref(git_dir)?.as_deref() == Some(ref_name),
        upstream: if kind == BranchType::Local {
            branch_upstream(git_dir, name)?
        } else {
            None
        },
    })
}

fn branch_ref_name(name: &str, kind: BranchType) -> String {
    match kind {
        BranchType::Local => local_branch_ref(name),
        BranchType::Remote => format!("refs/remotes/{}", name),
    }
}

fn local_branch_ref(name: &str) -> String {
    format!("refs/heads/{}", name)
}

fn short_branch_name(ref_name: &str, kind: BranchType) -> Option<&str> {
    match kind {
        BranchType::Local => ref_name.strip_prefix("refs/heads/"),
        BranchType::Remote => ref_name.strip_prefix("refs/remotes/"),
    }
}

fn branch_name_and_type(ref_name: &str) -> Option<(&str, BranchType)> {
    if let Some(name) = ref_name.strip_prefix("refs/heads/") {
        Some((name, BranchType::Local))
    } else {
        ref_name
            .strip_prefix("refs/remotes/")
            .map(|name| (name, BranchType::Remote))
    }
}

fn current_head_ref(git_dir: &Path) -> Result<Option<String>, MuonGitError> {
    let head = crate::refs::read_reference(git_dir, "HEAD")?;
    Ok(head
        .strip_prefix("ref: ")
        .map(|value| value.trim().to_string())
        .filter(|value| value.starts_with("refs/heads/")))
}

fn head_target_oid(git_dir: &Path) -> Result<OID, MuonGitError> {
    let head = crate::refs::read_reference(git_dir, "HEAD")?;
    if head.starts_with("ref: ") {
        crate::refs::resolve_reference(git_dir, "HEAD").map_err(|err| match err {
            MuonGitError::NotFound(_) => MuonGitError::UnbornBranch,
            other => other,
        })
    } else {
        OID::from_hex(head.trim())
    }
}

fn branch_section(name: &str) -> String {
    format!("branch.{}", name)
}

fn load_repo_config(git_dir: &Path) -> Result<Config, MuonGitError> {
    let config_path = git_dir.join("config");
    if config_path.exists() {
        Config::load(&config_path)
    } else {
        Ok(Config::with_path(&config_path))
    }
}

fn clear_branch_upstream(git_dir: &Path, name: &str) -> Result<(), MuonGitError> {
    let config_path = git_dir.join("config");
    if !config_path.exists() {
        return Ok(());
    }

    let mut config = Config::load(&config_path)?;
    let section = branch_section(name);
    config.unset(&section, "remote");
    config.unset(&section, "merge");
    config.save()
}

fn move_branch_upstream(git_dir: &Path, old_name: &str, new_name: &str) -> Result<(), MuonGitError> {
    let upstream = branch_upstream(git_dir, old_name)?;
    clear_branch_upstream(git_dir, old_name)?;
    if let Some(upstream) = upstream {
        set_branch_upstream(git_dir, new_name, Some(upstream))?;
    }
    Ok(())
}

fn reference_exists(git_dir: &Path, name: &str) -> bool {
    crate::refs::read_reference(git_dir, name).is_ok()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::PathBuf;
    use crate::refdb::packed_references;

    fn test_dir(name: &str) -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../tmp").join(name)
    }

    #[test]
    fn test_branch_create_lookup_list_and_upstream() {
        let tmp = test_dir("test_branch_create_lookup_list_and_upstream");
        let _ = fs::remove_dir_all(&tmp);
        let repo = Repository::init(tmp.to_str().unwrap(), false).unwrap();
        let git_dir = repo.git_dir();

        let main_oid = OID::from_hex("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa").unwrap();
        crate::refs::write_reference(git_dir, "refs/heads/main", &main_oid).unwrap();

        let branch = create_branch(git_dir, "feature", None, false).unwrap();
        assert_eq!(branch.name, "feature");
        assert_eq!(branch.reference_name, "refs/heads/feature");
        assert_eq!(branch.target.as_ref(), Some(&main_oid));
        assert!(!branch.is_head);

        set_branch_upstream(
            git_dir,
            "feature",
            Some(BranchUpstream {
                remote_name: "origin".into(),
                merge_ref: "refs/heads/main".into(),
            }),
        )
        .unwrap();
        let looked_up = lookup_branch(git_dir, "feature", BranchType::Local).unwrap();
        assert_eq!(
            looked_up.upstream,
            Some(BranchUpstream {
                remote_name: "origin".into(),
                merge_ref: "refs/heads/main".into(),
            })
        );

        let branches = list_branches(git_dir, Some(BranchType::Local)).unwrap();
        assert!(branches.iter().any(|entry| entry.name == "main" && entry.is_head));
        assert!(branches.iter().any(|entry| entry.name == "feature"));

        let _ = fs::remove_dir_all(&tmp);
    }

    #[test]
    fn test_branch_create_from_detached_head() {
        let tmp = test_dir("test_branch_create_from_detached_head");
        let _ = fs::remove_dir_all(&tmp);
        let repo = Repository::init(tmp.to_str().unwrap(), false).unwrap();
        let git_dir = repo.git_dir();

        let detached_oid = OID::from_hex("bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb").unwrap();
        fs::write(git_dir.join("HEAD"), format!("{}\n", detached_oid.hex())).unwrap();

        let branch = create_branch(git_dir, "detached-copy", None, false).unwrap();
        assert_eq!(branch.target.as_ref(), Some(&detached_oid));
        assert!(!branch.is_head);

        let _ = fs::remove_dir_all(&tmp);
    }

    #[test]
    fn test_branch_rename_updates_head_and_tracks_packed_refs() {
        let tmp = test_dir("test_branch_rename_updates_head_and_tracks_packed_refs");
        let _ = fs::remove_dir_all(&tmp);
        let repo = Repository::init(tmp.to_str().unwrap(), false).unwrap();
        let git_dir = repo.git_dir();

        let topic_oid = OID::from_hex("cccccccccccccccccccccccccccccccccccccccc").unwrap();
        fs::write(
            git_dir.join("packed-refs"),
            format!("{} refs/heads/topic\n", topic_oid.hex()),
        )
        .unwrap();
        crate::refs::write_symbolic_reference(git_dir, "HEAD", "refs/heads/topic").unwrap();
        set_branch_upstream(
            git_dir,
            "topic",
            Some(BranchUpstream {
                remote_name: "origin".into(),
                merge_ref: "refs/heads/main".into(),
            }),
        )
        .unwrap();

        let renamed = rename_branch(git_dir, "topic", "renamed", false).unwrap();
        assert_eq!(renamed.name, "renamed");
        assert_eq!(crate::refs::read_reference(git_dir, "HEAD").unwrap(), "ref: refs/heads/renamed");
        assert!(crate::refs::read_reference(git_dir, "refs/heads/topic").is_err());
        assert_eq!(renamed.target.as_ref(), Some(&topic_oid));
        assert_eq!(
            branch_upstream(git_dir, "renamed").unwrap(),
            Some(BranchUpstream {
                remote_name: "origin".into(),
                merge_ref: "refs/heads/main".into(),
            })
        );
        assert!(packed_references(git_dir).unwrap().get("refs/heads/topic").is_none());

        let _ = fs::remove_dir_all(&tmp);
    }

    #[test]
    fn test_branch_delete_rejects_checked_out_branch() {
        let tmp = test_dir("test_branch_delete_rejects_checked_out_branch");
        let _ = fs::remove_dir_all(&tmp);
        let repo = Repository::init(tmp.to_str().unwrap(), false).unwrap();
        let git_dir = repo.git_dir();

        let main_oid = OID::from_hex("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa").unwrap();
        crate::refs::write_reference(git_dir, "refs/heads/main", &main_oid).unwrap();

        let result = delete_branch(git_dir, "main", BranchType::Local);
        assert!(matches!(result, Err(MuonGitError::Conflict(_))));

        let _ = fs::remove_dir_all(&tmp);
    }
}
